pyogrio/_io.pyx

```diff
@@ -1238,6 +1240,35 @@ def ogr_read(
         field_data
     )
 
+
+cdef void pycapsule_array_stream_deleter(object stream_capsule) noexcept:
+    cdef ArrowArrayStream* stream = <ArrowArrayStream*>PyCapsule_GetPointer(
```

Comment:
`PyCapsule_GetPointer` sets an exception if the capsule name doesn't match (which should probably not happen, but is not impossible if someone messed with the object). This could conflict with the `noexcept` directive (also, destructors should not set exceptions because they will leak in unrelated code).

So I think you want to catch and log any exception here, however unlikely.

Reply:
I am not sure this is possible? The PyCapsule_Destructor can't actually raise errors (it's called from `capsule_dealloc` where errors are not checked).
When changing this, cython gives `"Cannot assign type 'void (*)(object) except *' to 'PyCapsule_Destructor'. Exception values are incompatible. Suggest adding 'noexcept' to type 'void (object) except *'."`

Or do you mean to print the message given that this cannot raise an error? (but I think that might already happen in that case? At least in other cases in pyarrow where we missed some correct error handling annotation in cython, you get messages like "exception ignored in ..." printed)

(this part is copied almost verbatim from pyarrow, so we should update that there as well)

Reply:
Oh, you're right, `noexcept` is actually part of the function signature. So we should keep it indeed.

But the question remains: what happens if `PyCapsule_GetPointer` raises an exception here? Does Cython catch the exception print a nice "unraisable error" message? Or does it let the exception stay and leak into unrelated code?

This could be checked by misspelling the capsule name and see what happens.

Reply:
Yes, as I expected, you get such a "exception ignored in" message to stdout (or stderr?). With changing the name inside the destructor so this always fails:

```
In [1]: import pyogrio

In [2]: with pyogrio.raw.open_arrow("test.geojson") as (meta, reader):
   ...:     pass
   ...: 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
ValueError: PyCapsule_GetPointer called with incorrect name
Exception ignored in: 'pyogrio._io.pycapsule_array_stream_deleter'
Traceback (most recent call last):
  File "/home/joris/conda/envs/geo-dev/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
ValueError: PyCapsule_GetPointer called with incorrect name
```

And pytest shows warns about that as unraisable exceptions:
```
pyogrio/tests/test_arrow.py: 36 warnings
  /home/joris/conda/envs/geo-dev/lib/python3.11/site-packages/_pytest/unraisableexception.py:80: PytestUnraisableExceptionWarning: Exception ignored in: 'pyogrio._io.pycapsule_array_stream_deleter'
  
  Traceback (most recent call last):
    File "/home/joris/conda/envs/geo-dev/lib/python3.11/contextlib.py", line 144, in __exit__
      next(self.gen)
  ValueError: PyCapsule_GetPointer called with incorrect name
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```

Reply:
(and since I'm curious, I posted [this](https://discuss.python.org/t/rationale-for-changing-a-capsules-name-or-destructor/48504))

Reply:
Thanks for checking! This can be resolved then.